### PR TITLE
Improve forbidden domain handling

### PR DESCRIPTION
**Pull request by Gallaecio**
https://github.com/zytedata/zyte-spider-templates/pull/13#discussion_r1386755312 (now with tests! :tada:)

To do:
- [x] Update after https://github.com/scrapy-plugins/scrapy-zyte-api/pull/150.

### CODE REVIEW

File: scrapy_zyte_api/_middlewares.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        if not self._total_start_request_count:
+            return
+        if self._failed_start_request_count < self._total_start_request_count:
+            return
+        logger.error(
+            "Stopping the spider, all start request failed because they "
+            "were pointing to a domain forbidden by Zyte API."
+        )
+        self._crawler.engine.close_spider(
+            self._crawler.spider, "failed-forbidden-domain"
```
Comment (Gallaecio) [inline scrapy_zyte_api/_middlewares.py:130]: To do: Use snake_case for consistency with upstream.

File: scrapy_zyte_api/_middlewares.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    def process_start_requests(self, start_requests, spider):
+        count = 0
+        for request in start_requests:
+            request.meta["is_start_request"] = True
+            yield request
+            count += 1
+        self._send_signal(_start_requests_processed, count=count)
+
+
+class ForbiddenDomainDownloaderMiddleware:
```
Comment (BurnzZ) [inline scrapy_zyte_api/_middlewares.py:89]: What do you think about merging the logic to [ScrapyZyteAPIDownloaderMiddleware](https://github.com/scrapy-plugins/scrapy-zyte-api/blob/main/scrapy_zyte_api/_downloader_middleware.py#L10)?

Comment (BurnzZ) [reply scrapy_zyte_api/_middlewares.py:89]: Yeah, I think we want it to be present at all times.

File: tests/test_providers.py [hunk 3]
```diff
@@ -22,6 +19,19 @@
...
+def create_scrapy_settings():
+    settings = _create_scrapy_settings(None)
+    for setting, value in SETTINGS.items():
+        if setting.endswith("_MIDDLEWARES"):
+            if settings[setting]:
+                settings[setting].update(value)
+            else:
+                settings[setting] = value
+        else:
+            settings[setting] = value
```
Comment (BurnzZ) [inline tests/test_providers.py:31]:
```suggestion
        if setting.endswith("_MIDDLEWARES") and settings[setting]:
            settings[setting].update(value)
        else:
            settings[setting] = value
```

File: tests/test_middlewares.py [hunk 4]
```diff
@@ -132,3 +132,130 @@ def parse(self, response):
...
+        crawler = get_crawler(TestSpider, settings_dict=settings)
+        await crawler.crawl()
+
+    assert crawler.stats.get_value("finish_reason") == "finished"
+
+
+@ensureDeferred
+async def test_forbidden_domain_with_partial_start_request_consumption():
+    """With concurrency lower than the number of start requests + 1, the code
+    path followed changes, because ``__total_start_request_count`` is not set
```
Comment (BurnzZ) [inline tests/test_middlewares.py:237]:
```suggestion
    path followed changes, because ``_total_start_request_count`` is not set
```

**Review by BurnzZ — APPROVED** [review]

**Comment by Gallaecio** [comment]
We can ignore the docs error, it’s solved in 2 other PRs already.

**Review by kmike — APPROVED** [review]

**Pull request merged**